datafusion/core/src/datasource/file_format/write/orchestration.rs

```diff
@@ -50,7 +50,7 @@ pub(crate) async fn serialize_rb_stream_to_object_store(
     mut data_rx: Receiver<RecordBatch>,
     serializer: Arc<dyn BatchSerializer>,
     mut writer: WriterType,
-) -> std::result::Result<(WriterType, u64), (WriterType, DataFusionError)> {
+) -> std::result::Result<(WriterType, u64), (Option<WriterType>, DataFusionError)> {
```

Comment:
not for this PR, but this signature is getting pretty gnarly. Maybe it is time to try and encapsulate some of this logic into structs 🤔

Reply:
I think the rationale would be hard to discern only from the code. Could you please update the documentation to reflect the change of what is returned on error?

Specifically, I think it is that if there was an error on write, the writer is dropped so we don't accidentally write to it or try and close it. For any error not involving the writer, the writer is returned as well

cc @metesynnada  as I remember at some point maybe you had a usecase for calling `shutdown` on a writer after error 🤔 I may be misremembering

Reply:
Makes sense, I will update the docs

Reply:
Filed https://github.com/apache/datafusion/issues/11443 to track cleaning up the megatype